sky/data/storage.py

```diff
@@ -1871,6 +1904,542 @@ def _delete_gcs_bucket(self, bucket_name: str) -> bool:
+            pattern = r'^[a-z0-9][-a-z0-9]*[a-z0-9]$'
+            if not re.match(pattern, name):
+                _raise_no_traceback_name_error(
+                    f'Invalid store name: name {name} can consist only of '
+                    'lowercase letters, numbers, and hyphens (-). '
+                    'It must begin and end with a letter or number.')
+
+            # Check for two adjacent hyphens
+            if '--' in name:
+                _raise_no_traceback_name_error(
+                    f'Invalid store name: name {name} must not contain '
+                    'two adjacent hyphens.')
+
+        else:
+            _raise_no_traceback_name_error('Store name must be specified.')
+        return name
+
+    def initialize(self):
+        """Initializes the AZ Container object on the cloud.
+
+        Initialization involves fetching bucket if exists, or creating it if
+        it does not. Also, it checks for the existance of the storage account
+        and resource group if provided by the user. If not provided, those are
+        created with a default naming convention.
+
+        Raises:
+            StorageBucketCreateError: If bucket creation fails
+            StorageBucketGetError: If fetching existing bucket fails
+            StorageInitError: If general initialization fails.
+        """
+        self.storage_client = data_utils.create_az_client('storage')
+        self.resource_client = data_utils.create_az_client('resource')
+        # Using externally created storage
+        if isinstance(self.source, str) and self.source.startswith('az://'):
+            bucket_name, _, storage_account_name = data_utils.split_az_path(
+                self.source)
+            assert self.name == bucket_name
+            self.storage_account_name = storage_account_name
+            self.resource_group_name = data_utils.get_az_resource_group(
+                storage_account_name)
+        # Creating new resource group and storage account
+        else:
+            # If resource group or storage account names are not provided from
+            # config, then use default names.
+            self.storage_account_name = skypilot_config.get_nested(
+                ('azure', 'storage_account'),
+                f'sky{self.region}{common_utils.get_user_hash()}')
+            self.resource_group_name = skypilot_config.get_nested(
+                ('azure', 'resource_group'),
+                f'sky{common_utils.get_user_hash()}')
+            # check if the resource group with the given name already exists.
+            try:
+                self.resource_client.resource_groups.get(
+                    self.resource_group_name)
+            except azure.exceptions().ResourceNotFoundError as e:
+                if 'Code: ResourceGroupNotFound' in e.message:
+                    self.resource_client.resource_groups.create_or_update(
+                        self.resource_group_name, {'location': self.region})
```

Comment:
1. Removed the check for resource group, and we are solely running `create_or_update`. This is done at [80e1bce](https://github.com/skypilot-org/skypilot/pull/3032/commits/80e1bceaceb8bc981d93c0f9b923a6d5d87f83d9)
2. Have some questions regards to using `get_azure_sdk_function`.
i. So for every function calls that involves azure sdk client, which are currently spread out in `cloud_stores.py`, `storage.py`, and `data_utils.py` for this PR, you want those to retrieve the function through `get_azure_sdk_function` first instead of directly calling the function immediately, right? My first impression of this was that it seems to increase the complexity, but I assume I'm lacking understanding of the benefit of such approach. Do you mind helping to elaborate a little on this one? I guess I didn't fully understand what you meant by 'better handle the api inconsistency in the azure core'.
ii. Seems like there's a redundant `get_azure_sdk_function` in both `sky/provision/azure/instance.py` and `/home/gcpuser/skypilot/sky/skylet/providers/azure/config.py`. Now this function needs to be used by 3 additional modules. What would be the best location to place this function so all the modules can access it? I'm thinking of `common_utils.py` perhaps? What do you think?

Reply:
1. However, iirc, if we `sky launch` on an existing Azure cluster, it will directly use this `create_or_update` function instead of checking if it exists first. Could you look at the differences b/t your usage and that part of the code and investigate why it does not works for your usage? A suggestion: Maybe try specify the locations in the parameters, like here.
https://github.com/skypilot-org/skypilot/blob/eedeed468626be1e831589543f90632477adfc41/sky/skylet/providers/azure/config.py#L71
2. i. This is mainly due to an API change from Azure core. They renamed a bunch of functions from `begin_xxx` to `xxx`, e.g. `begin_create_or_update ==> create_or_update`. Using the `get_azure_sdk_function` would make our code more compatible with older versions of azure core on the user's machine. Though I'm not sure of the exact timestamp they made this change and maybe it is a while ago and we don't really need to do this. cc @Michaelvll for a look here.
2. ii. The file `skypilot/sky/skylet/providers/azure/config.py` will be removed once the new provisioner of Azure is added to the system. I guess we could reuse the one from `sky/provision/azure/instance.py`. Also cc Zhanghao for a look here 🤔